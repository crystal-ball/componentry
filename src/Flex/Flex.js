import { useTheme } from '../Theme/Theme'
import { element } from '../factories/element'

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export function Flex(props) {
  /** @type {{ align: string, direction: string, inline: boolean, justify: string, wrap: string }} */
  const { align, direction, inline, justify, wrap, ...rest } = {
    ...useTheme('Flex'),
    ...props,
  }

  return element({
    componentCx: {
      // ‚ÑπÔ∏è Flex is a helper component so it doesn't have a üÖ≤-flex class
      'd-flex': !inline,
      'd-inline-flex': inline,
      // column, column-reverse, row, row-reverse
      [`flex-${direction}`]: direction,
      // wrap, wrap-reverse, nowrap
      [`flex-${wrap}`]: wrap,
      // start, end, center, baseline, stretch
      [`align-items-${align}`]: align,
      // start, end, center, around, between, evenly
      [`justify-content-${justify}`]: justify,
    },
    ...rest,
  })
}
Flex.displayName = 'Flex'
