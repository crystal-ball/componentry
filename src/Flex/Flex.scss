/* ~componentry/src/Flex/Flex.scss */

/// Componentry Flex styles that don't map to a Tailwind utility style are
/// included as a component style to ensure they're available for users that
/// switch to Tailwind for their utility styles

@each $breakpoint in map-keys(theme-get('breakpoints')) {
  @include media-breakpoint-up($breakpoint) {
    $infix: breakpoint-infix($breakpoint, theme-get('breakpoints'));

    // Provide "flex-column" and "flex-column-reverse" classes to map to Flex's
    // direction prop values.
    // (Tailwind provides "flex-col" and "flex-col-reverse", though short these
    // don't map to the CSS styles and Componentry prioritizes consistent
    // intuition of values over brevity)

    .#{$infix}flex-column {
      flex-direction: column !important;
    }

    .#{$infix}flex-column-reverse {
      flex-direction: column-reverse !important;
    }
  }
}
