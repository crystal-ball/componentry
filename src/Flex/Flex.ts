import React from 'react'
import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../utils/types'
import { element } from '../utils/element-creator'

interface FlexProps extends ComponentBaseProps<'div'> {
  /** Sets an `align-items` flex style */
  align?: 'start' | 'center' | 'end' | 'stretch' | 'baseline'
  /** sets a `flex-direction` flex style */
  direction?: 'row' | 'column' | 'row-reverse' | 'column-reverse'
  /** Switches between display between an inline and block element */
  inline?: boolean
  /** Sets a `justify-content` flex style */
  justify?: 'start' | 'center' | 'end' | 'between' | 'around' | 'evenly'
  /** Sets a `flex-wrap` flex style */
  wrap?: 'wrap' | 'nowrap' | 'wrap-reverse'
}

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export const Flex: React.FC<FlexProps> = (props) => {
  const { align, direction, inline, justify, wrap, ...rest } = {
    ...useTheme<FlexProps>('Flex'),
    ...props,
  }

  return element('Flex', {
    componentCx: {
      'd-flex': !inline,
      'd-inline-flex': inline,
      [`flex-${direction}`]: direction,
      [`flex-${wrap}`]: wrap,
      [`align-items-${align}`]: align,
      [`justify-content-${justify}`]: justify,
    },
    ...rest,
  })
}
Flex.displayName = 'Flex'
