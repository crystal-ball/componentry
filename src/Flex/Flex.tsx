import React from 'react'
import { useTheme } from '../Theme/Theme'
import { element } from '../factories/element'
import { BaseProps } from '../utils/base-types'

interface FlexProps
  extends BaseProps,
    Omit<React.ComponentPropsWithoutRef<'div'>, 'className'> {
  /** Sets an `align-items` flex style */
  align?: 'start' | 'center' | 'end' | 'stretch' | 'baseline'
  /** sets a `flex-direction` flex style */
  direction?: 'row' | 'column' | 'row-reverse' | 'column-reverse'
  /** Switches between display between an inline and block element */
  inline?: boolean
  /** Sets a `justify-content` flex style */
  justify?: 'start' | 'center' | 'end' | 'between' | 'around' | 'evenly'
  /** Sets a `flex-wrap` flex style */
  wrap?: 'wrap' | 'nowrap' | 'wrap-reverse'
}

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export const Flex: React.FC<FlexProps> = (props) => {
  const { align, direction, inline = false, justify, wrap, ...rest } = {
    ...useTheme<FlexProps>('Flex'),
    ...props,
  }

  return element({
    componentCx: {
      // ‚ÑπÔ∏è Flex is a helper component so it doesn't have a üÖ≤-flex class
      'd-flex': !inline,
      'd-inline-flex': inline,
      // column, column-reverse, row, row-reverse
      [`flex-${direction}`]: direction,
      // wrap, wrap-reverse, nowrap
      [`flex-${wrap}`]: wrap,
      // start, end, center, baseline, stretch
      [`align-items-${align}`]: align,
      // start, end, center, around, between, evenly
      [`justify-content-${justify}`]: justify,
    },
    ...rest,
  })
}
Flex.displayName = 'Flex'
