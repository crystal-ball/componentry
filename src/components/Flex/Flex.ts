import React from 'react'
import { useTheme } from '../Theme/Theme'
import { ComponentBaseProps } from '../../utils/types'
import { element } from '../../utils/element-creator'

export interface FlexProps extends ComponentBaseProps<'div'> {
  /** Sets an `align-items` style */
  align?: 'start' | 'end' | 'center' | 'baseline' | 'stretch'
  /** Sets a `flex-direction` flex style */
  direction?: 'column' | 'column-reverse' | 'row-reverse' | 'row'
  /** Switches between display between an inline and block element */
  inline?: boolean
  /** Sets a `justify-content` style */
  justify?: 'start' | 'end' | 'center' | 'between' | 'around' | 'evenly'
  /** Sets a `flex-wrap` flex style */
  wrap?: 'wrap' | 'nowrap' | 'wrap-reverse'
}

/**
 * [Flex component üìù](https://componentry.design/components/flex)
 */
export const Flex: React.FC<FlexProps> = (props) => {
  const { align, direction, inline, justify, wrap, ...rest } = {
    ...useTheme<FlexProps>('Flex'),
    ...props,
  }
  // Tailwind uses a flex-col class but direction="col" is super wonky
  // => so props use "column" and we replace with "col"
  const computedDirection = direction?.replace('column', 'col')

  return element({
    componentCx: {
      'flex': !inline,
      'inline-flex': inline,
      [`flex-${computedDirection}`]: computedDirection,
      [`flex-${wrap}`]: wrap,
    },
    alignItems: align,
    justifyContent: justify,
    ...rest,
  })
}
Flex.displayName = 'Flex'
