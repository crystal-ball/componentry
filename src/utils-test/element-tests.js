/* eslint-disable import/no-extraneous-dependencies */
import React from 'react'
import { mount } from 'enzyme'

import dt from './dt'

/**
 * Library default component requirment test suite.
 *
 * IMPORTANT NOTES
 * - Enzyme `find()` will find **EVERY** instance of a prop, this makes it difficult
 *   to assert on find lengths, because it will be variable for components with
 *   HOCs which is why we use `toBeTruthy` to only check that some instance of the
 *   expected element attr exists
 */

export default (TestComponent, enzymeOptions = {}) => {
  /*
   * Test that any html tag or component can be passed through the 'as' prop to
   * override the parent element the component renders
   */
  test('should render as specified html element or component', () => {
    // Create a component to validate that the TestComponent returns.
    const TestAs = () => <section />
    const wrapper = mount(<TestComponent as={TestAs} />, enzymeOptions)
    expect(wrapper.find(TestAs).length).toBeTruthy()
  })

  /*
   * All components should spread passed props on the parent container for max
   * customization
   */
  test('should pass through classnames and attributes', () => {
    const wrapper = mount(
      <TestComponent className="test-custom" data-test="test-custom" />,
      enzymeOptions,
    )
    expect(wrapper.find('.test-custom').length).toBeTruthy()
    expect(wrapper.find(dt('test-custom')).length).toBeTruthy()
  })

  // TODO: should render passed children
}
