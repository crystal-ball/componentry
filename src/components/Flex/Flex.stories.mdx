import { ArgsTable, Meta } from '@storybook/addon-docs'
import { BADGE } from '@geometricpanda/storybook-addon-badges'

import { Flex } from './Flex'

<Meta
  title='Components/Flex'
  component={Flex}
  parameters={{ badges: [BADGE.EXPERIMENTAL] }}
/>

# Flex

Flex is a convenience component for creating `display: flex` containers.

<ArgsTable of={Flex} />

## Block containers

Flex components create block level flex containers:

<Flex>
  {[1, 2, 3].map((key) => (
    <div key={key} className='border'>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Inline containers

Create inline level flex containers using the `inline` prop:

<Flex inline>
  {[1, 2].map((key) => (
    <div key={key} className='border'>
      Flex content #{key}
    </div>
  ))}
</Flex>
<Flex inline>
  {[1, 2].map((key) => (
    <div key={key} className='border'>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Flex direction

Override the default direction of `row` with the `direction` prop:

<Flex direction='row-reverse'>
  {[1, 2, 3].map((key) => (
    <div key={key} className='border'>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Justify and align

Set the flex items justification and alignment with the `justify` and `align` props:

<Flex justify='center' align='center'>
  {[1, 2, 3].map((key) => (
    <div key={key} className='border'>
      Flex content #{key}
    </div>
  ))}
</Flex>

## Flex wrap

Override the flex wrap behavior with the `wrap` props:

<Flex wrap='wrap-reverse'>
  {Array(12)
    .fill(0)
    .map((_, key) => (
      // eslint-disable-next-line react/no-array-index-key
      <div key={key} className='border'>
        Flex content #{key}
      </div>
    ))}
</Flex>
