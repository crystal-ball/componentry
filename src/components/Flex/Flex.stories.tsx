import type { Meta, StoryObj } from '@storybook/react'

import { Flex } from './Flex'

const meta: Meta<typeof Flex> = {
  component: Flex,
}

export default meta
type Story = StoryObj<typeof Flex>

export const Primary: Story = {
  args: {
    backgroundColor: 'primary-200',
    className: 'rounded',
    children: [1, 2, 3].map((num) => (
      <div
        key={num}
        className='bg-primary-600 text-inverse font-mono font-bold px-6 py-4 rounded'
      >
        #{num > 9 ? num : `0${num}`}
      </div>
    )),
  },
}
