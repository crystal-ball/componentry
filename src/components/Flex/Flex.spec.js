import { render, screen } from '@testing-library/react'

import { elementTests } from '../../test/element-tests'
import { Flex } from './Flex'

describe('<Flex />', () => {
  elementTests(Flex)

  it('Renders class flex by default', () => {
    render(<Flex>Content</Flex>)

    expect(screen.getByText('Content')).toHaveClass('flex')
  })

  it('When display is passed, then class flex is overridden', () => {
    render(<Flex display='inline-flex'>Content</Flex>)

    expect(screen.getByText('Content')).toHaveClass('inline-flex')
  })

  it('When modifier props are passed, then the expanded classNames are rendered', () => {
    render(
      <Flex align='start' direction='column' justify='start' wrap='wrap'>
        Content
      </Flex>,
    )

    expect(screen.getByText('Content')).toHaveClass(
      'flex flex-col items-start flex-wrap justify-start',
    )
  })
})

// Snapshots
// ---------------------------------------------------------------------------
describe('<Flex /> snapshots', () => {
  it('renders correctly', () => {
    render(<Flex data-testid='flex'>Flex content</Flex>)

    expect(screen.getByTestId('flex')).toMatchSnapshot()
  })
})
