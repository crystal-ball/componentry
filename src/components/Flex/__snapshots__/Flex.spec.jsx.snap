// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Flex /> snapshots > renders correctly 1`] = `
<div
  class="flex"
  data-testid="flex"
>
  Flex content
</div>
`;
